### ISSUE TITLE
Strange generator behaviour

### ISSUE THREAD

**Issue by emilbols** [DL4Jets/DeepJetCore#28]
I have noticed recently that all models, i train with the deepjet generator seems to optimize worse than they did in the past. It seems that this happened when i upgraded to keras 2.2.4 and tf 1.13.1, but how that should have an effect on the generator, im not sure. I tried to compare model.fit_generator with just model.fit, where i load in the dataset in a manner like this.

https://github.com/emilbols/DeepJetCore/blob/1ac93801bcf0a126763041007beb78f0bbf5c5c4/training/training_base.py#L544-L556

Here is the batch loss for deepFlavour when loading the dataset in memory:
![NoGen_lossperbatch](https://user-images.githubusercontent.com/33452471/64165528-9175ea00-ce45-11e9-897d-a86a89019369.png)

But when i use the generator with the same model and dataCollection, i get this batch loss:
![Gen_lossperbatch](https://user-images.githubusercontent.com/33452471/64165554-9dfa4280-ce45-11e9-9d40-084e7319fc08.png)

If you compare the epoch loss, it seems to affect the optimization. I definitely did not see this behaviour in the past:
![loss](https://user-images.githubusercontent.com/33452471/64165574-a5215080-ce45-11e9-8505-040c75027f73.png)

Looking at the batches that are fed before you have these explosions, i did not see anything obvious, no NaNs or huge values as far as i could tell. 

Any ideas?

**Comment by jkiesele** [comment]
Hi Emil,

that is indeed very weird. In particular because in the end the data is read in a similar manner in both cases.
Do you have a chance to revert back to older keras/tf versions and see if the behaviour persists?

You can also check if it has something to do with the threaded read in, by replacing the readIn_async here:
https://github.com/DL4Jets/DeepJetCore/blob/master/DataCollection.py#L821
by a simple readIn, and change wasasync=True to False here:
https://github.com/DL4Jets/DeepJetCore/blob/master/DataCollection.py#L848

This will be slower, but might be a good check if it has anything to do with the threading.
Unfortunately, I don't have better suggestions.

**Comment by emilbols** [comment]
since readIn doesn't take the ramdisk and shuffle argument, should i just leave them empty in L821 or should i pass them to the readIn_async function called within readIn?
https://github.com/DL4Jets/DeepJetCore/blob/master/TrainData.py#L581 

also should this be kept?
https://github.com/DL4Jets/DeepJetCore/blob/master/DataCollection.py#L834-L835

also it seems that you get stuck here:
https://github.com/DL4Jets/DeepJetCore/blob/c988bd135fb9f39a96c6a398bcdd8380daec751d/TrainData.py#L494-L496

when running with async = false. I assume because it appears to be left empty
https://github.com/DL4Jets/DeepJetCore/blob/c988bd135fb9f39a96c6a398bcdd8380daec751d/TrainData.py#L451

unlike when async true
https://github.com/DL4Jets/DeepJetCore/blob/c988bd135fb9f39a96c6a398bcdd8380daec751d/TrainData.py#L416-L436

**Comment by emilbols** [comment]
So, i cross checked a  number of things, including the threading, whether labels are somehow mismatched, whether it is connected to the batchnorm layer. everything looked fine.

Then I tried to backport to keras 2.0.0 which we used in the past, and the issue appears to be gone. I tried out a number of keras version, and it seems 2.0.5 is the latest one without this issue, as you can see here:

![KERAS_2_0_5_gpu_DF_lossperbatch](https://user-images.githubusercontent.com/33452471/64336917-696dbe80-cfde-11e9-82a1-449c354cc91a.png)

It also seems that these spikes happen more and more frequently as you move to newer keras version, but that might just be a coincidence, im not sure.

looking at the keras changelog they did apparently change some stuff with the fit generator method going into keras 2.0.6, perhaps it is connected

https://github.com/keras-team/keras/releases?after=2.1.3

anyways, i will start a new df training today with 2.0.5, just to check it converges like it used to in the past

**Comment by jkiesele** [comment]
Thanks a lot for these checks!
Does keras 2.0.5 still work with tensorflow 1.13?

**Comment by emilbols** [comment]
I believe so, im using 1.14, so surely 1.13 should be possible.

**Comment by jkiesele** [comment]
OK, I am closing this one since it is "fixed", but leave the other one (#30 ) open
